### PR TITLE
chore: Mitigate auction front-running (VE)

### PR DESCRIPTION
**Pull request by samricotta**
Tutorial for implementation of Vote Extensions. Uses https://github.com/fatal-fruit/abci-workshop. This will be moved across to the `sdk-tutorials` repo next week but for now will point to the original repo that was used for the workshop

Closes: https://github.com/cosmos/cosmos-sdk-docs/issues/90

### CODE REVIEW

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 1]
```diff
@@ -0,0 +1,431 @@
+# Mitigate Auction Front-Running
+
+## What are Vote extensions?
+
+Vote extensions is arbitrary information which can be inserted into a block. This feature is part of ABCI 2.0, which is available for use in the SDK 0.50 release and part of the 0.38 CometBFT release. 
+
+More information about vote extensions can be seen [here](../../../build/building-apps/vote-extensions).
```
Comment (facundomedica) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:7]: nit: link might be dead

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+## Context
+
+For this tutorial we are using an example of an application that is mitigating auction front-running. 
+
+Please checkout https://github.com/fatal-fruit/abci-workshop to see the example.
+
+
+## Simulation of Auction Front-Running
+
+Please checkout `part 1` of the example and then run:
```
Comment (facundomedica) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:18]:
```suggestion
Please checkout the branch `part-1` of the example and then run:
```

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+./build/cosmappd keys list --home $HOME/.cosmappd --keyring backend test
+```
+
+Match the two addresses from the query and keys. The owner should be `val1`
+
+
+## Mitigating Front-running with Vote Extensions Steps
+
+
+To initiate the process, we introduce new types within the `abci/types.go` file. Carefully consider the specific requirements for each Type, bearing in mind that we utilise our own mempool for transaction ordering.
```
Comment (facundomedica) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:64]:
```suggestion
To initiate the process, we introduce new types within the `abci/types.go` file. Carefully consider the specific requirements for each type, bearing in mind that we utilise our own mempool for transaction ordering.
```

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+```
+
+
+Within `processVoteExtensions` method, all of the vote extensions are retrieved from the last commit and are turned into a special transaction. The special transaction is appended to the proposal.
+
+```go
+func (h *PrepareProposalHandler) PrepareProposalHandler() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+
+      if req.Height > 2 {
```
[the remark this answers is not in the document]
Comment (samricotta) [reply docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:211]: Will this still work though with the old rc of the sdk? I have to update the whole workshop so we are on the final release of 0.50

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+
+Please checkout https://github.com/fatal-fruit/abci-workshop to see the example.
+
+
+## Simulation of Auction Front-Running
+
+Please checkout the branch `part-1` of the example and then run:
+
+```shell
+./scripts/single_node/setup.sh
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:21]: short explanation of what this script does would be useful. asking people to run scripts blindly is never good

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+```
+
+In another window run:
+
+```shell
+./scripts/single_node/frontrun.sh
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:27]: ditto

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+```
+
+Here we can see alice attempts to reserve the namespace `bob.cosmos`
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:30]: is this related with the above scripts?

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+
+To verify the occurrence of front-running, navigate to the logs and search for instances of `Found a Bid to Snipe`. This specific message indicates the desired behavior, demonstrating that front-running has occurred.
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:32]: can we see an example of how it would look like? what do i do if i dont see it?

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+
+Below we query the namespace `bob.cosmos` to see if alice managed to successfully reserve `bob.cosmos`
+
+```shell 
+./build/cosmappd query ns whois bob.cosmos
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:37]: how do i get this app built? didnt see it in the walk through or prerequisites

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+  - amount: "2000"
+    denom: uatom
+  name: bob.cosmos
+  owner: cosmos185gc7c296w0xjlq9kjdt6gghnqvdmyckv64e7a
+  resolve_address: cosmos185gc7c296w0xjlq9kjdt6gghnqvdmyckv64e7a
+```
+
+The owner address corresponds to Alice (`val1` in the keyring)
+
+## Mitigating Front-running with Vote Extensions Steps
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:54]:
```suggestion
## Mitigating Front-running with Vote Extensions
```

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+To initiate the process, we introduce new types within the `abci/types.go` file. Carefully consider the specific requirements for each type, bearing in mind that we utilise our own mempool for transaction ordering.
+
+```go
+package abci
+
+import (
+	//import the necessary files
+)
+
+type PrepareProposalHandler struct {
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:66]: am i suppose to copy this over? 

if this is an advanced tutorial we should explain the structs so people understand why its needed

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+type VoteExtHandler struct {
+	logger       log.Logger
+	currentBlock int64
+	mempool      *mempool.ThresholdMempool
+	cdc          codec.Codec
+}
+
+type InjectedVoteExt struct {
+	VoteExtSigner []byte
+	Bids          [][]byte
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:91]: are bids encoding transactions or ?

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+	logger      log.Logger
+	txConfig    client.TxConfig
+	cdc         codec.Codec
+	mempool     *mempool.ThresholdMempool
+	txProvider  provider.TxProvider
+	keyname     string
+	runProvider bool
+}
+
+type ProcessProposalHandler struct {
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:76]: im doing vote extensions why do i need process and prepare handlers?

**Comment by tac0turtle** [comment]
im kind of confused by this pr, its unclear if its a tutorial or a walkthrough of something already existing? Who is the audience for this tutorial as well? Advanced users or beginners?

**Comment by facundomedica** [comment]
> Who is the audience for this tutorial as well? Advanced users or beginners?

I think the audience is more advanced users, because to be able to use vote extensions you need to already know lots of other things prior to this

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 14]
```diff
@@ -0,0 +1,500 @@
...
+
+For this tutorial we are using an example of an application that is mitigating auction front-running. 
+
+Front-running is a potential issue in blockchain systems where a participant gains an advantage by being able to perform actions or transactions ahead of others. In the context of this application, front-running can occur when a validator, who is also a proposer, spots a bid in the mempool and replaces it with their own higher bid. This is also known as "bid sniping".
+
+Please checkout https://github.com/fatal-fruit/abci-workshop to see the example.
+
+## Simulation of Auction Front-Running
+
+To verify an occurrence of front-running, you can navigate to the logs and search for instances of `💨 :: Found a Bid to Snipe`. This specific log message indicates that the validator has identified a bid to potentially front-run.
```
Comment (facundomedica) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:27]: This line maybe needs to be removed or moved a few paragraphs below, as it's something you are later instructed to do

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 15]
```diff
@@ -0,0 +1,500 @@
...
+
+## What are Vote extensions?
+
+Vote extensions is arbitrary information which can be inserted into a block. This feature is part of ABCI 2.0, which is available for use in the SDK 0.50 release and part of the 0.38 CometBFT release. 
+
+More information about vote extensions can be seen [here](../../../build/abci/03-vote-extensions.md).
+
+## Context
+
+For this tutorial we are using an example of an application that is mitigating auction front-running.
```
Comment (facundomedica) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:19]: Can we expand a little bit on what the example application does? Explain beforehand that the application is an auctioning of namespaces, just a couple of lines is fine.

File: docs/user/tutorials/vote-extensions/mitigate-frontrunning.md [hunk 16]
```diff
@@ -0,0 +1,500 @@
...
+
+More information about vote extensions can be seen [here](../../../build/abci/03-vote-extensions.md).
+
+## Context
+
+For this tutorial we are using an example of an application that is mitigating auction front-running. 
+
+Front-running is a potential issue in blockchain systems where a participant gains an advantage by being able to perform actions or transactions ahead of others. In the context of this application, front-running can occur when a validator, who is also a proposer, spots a bid in the mempool and replaces it with their own higher bid. This is also known as "bid sniping".
+
+Please checkout https://github.com/fatal-fruit/abci-workshop to see the example.
```
Comment (tac0turtle) [inline docs/user/tutorials/vote-extensions/mitigate-frontrunning.md:23]: can we get this migrated to the tutorial repo? we usually wouldnt allow this sort of linking for something we will end up hosting in the tutorial repo get merged as it will cause confusion later on. 

we should also mention that the tutorial is in that repo and this walkthrough is assuming you have gone through the tutorial.
